modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/Outbox.java

```diff
@@ -0,0 +1,277 @@
+        if (cancelled)
+            return;
+
+        try {
+            perNodeBuffers.values().forEach(Buffer::cancel);
+            input().cancel();
+        }
+        finally {
+            cancelled = true;
+            close();
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override public void target(Sink<T> sink) {
+        throw new AssertionError();
+    }
+
+    /** {@inheritDoc} */
+    @Override public Sink<T> sink(int idx) {
+        if (idx != 0)
+            throw new IndexOutOfBoundsException();
+
+        return this;
+    }
+
+    /** {@inheritDoc} */
+    @Override public boolean push(T row) {
+        List<UUID> nodes = function.destination(row);
+
+        if (F.isEmpty(nodes))
+            return true;
+
+        List<Buffer> buffers = new ArrayList<>(nodes.size());
+
+        for (UUID node : nodes) {
+            Buffer dest = perNodeBuffers.computeIfAbsent(node, this::createBuffer);
+
+            if (!dest.ready())
+                return false;
+
+            buffers.add(dest);
+        }
+
+        for (Buffer dest : buffers)
+            dest.add(row);
+
+        return true;
+    }
+
+    /** {@inheritDoc} */
+    @Override public void end() {
+        for (UUID node : function.targets())
+            perNodeBuffers.computeIfAbsent(node, this::createBuffer).end();
+
+        close();
+    }
+
+    /** {@inheritDoc} */
+    @Override public void close() {
+        context().mailboxRegistry().unregister(this);
+    }
+
+    /** */
+    private void sendBatch(UUID nodeId, int batchId, List<?> rows) {
+        context().exchange().sendBatch(this, nodeId, queryId(), targetFragmentId, exchangeId, batchId, rows);
+    }
+
+    /** */
+    private void sendCancel(UUID nodeId, int batchId) {
+        context().exchange().cancel(this, nodeId, queryId(), targetFragmentId, exchangeId, batchId);
+    }
+
+    /** */
+    private Buffer createBuffer(UUID nodeId) {
+        return new Buffer(nodeId, this);
+    }
+
+    /** */
+    private static final class Buffer {
+        /** */
+        private final Outbox<?> owner;
+
+        /** */
+        private final UUID nodeId;
+
+        /** */
+        private int hwm = -1;
+
+        /** */
+        private int lwm = -1;
+
+        /** */
+        private List<Object> curr;
+
+        /** */
+        private Buffer(UUID nodeId, Outbox<?> owner) {
+            this.nodeId = nodeId;
+            this.owner = owner;
+
+            curr = new ArrayList<>(ExchangeService.BATCH_SIZE + 1); // extra space for end marker;
+        }
+
+        /**
+         * Adds a row to current batch.
+         *
+         * @param row Row.
+         */
+        public void add(Object row) {
+            assert ready();
+
+            if (curr.size() == ExchangeService.BATCH_SIZE) {
+                owner.sendBatch(nodeId, ++hwm, curr);
+                curr = new ArrayList<>(ExchangeService.BATCH_SIZE + 1); // extra space for end marker;
+            }
+
+            curr.add(row);
```

Comment:
There is an another example when you cannot send a batch because there are too many unacknowledged ones. Such batch (what you cannot send) is used as an indicator to stop execution (return false from push() method). Also it's good to postpone sending (until a next row is pushed) because some of acks may be received in the middle, so, you wasn't able to send it when it became full but able to send at the time a next row is being sent. Such approach a bit simplified code reducing some checks.

Reply:
Ok